zeroconf/__init__.py

```diff
@@ -2865,20 +2868,117 @@ def __init__(
+        """Start a thread with a running event loop."""
+
+        def _run_loop() -> None:
+            loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(loop)
+            self._start_async(loop)
+            assert self._loop is not None
+            self._loop.run_forever()
+
+        self._loop_thread_ready = threading.Event()
+        self._loop_thread = threading.Thread(target=_run_loop, daemon=True)
+        self._loop_thread.start()
+        self._loop_thread_ready.wait()
+
+    async def async_setup(self) -> None:
+        """Set up the instance."""
+        assert self._loop is not None
+        await self._async_create_endpoints()
+        self._cache_cleanup_task = self._loop.create_task(self.async_cache_cleanup())
+        assert self._running_event is not None
+        self._running_event.set()
+        if self._loop_thread_ready:
+            self._loop_thread_ready.set()
+
+    async def _async_create_endpoints(self) -> None:
+        """Create endpoints to send and receive."""
+        assert self._loop is not None
+        loop = self._loop
+        reader_sockets = []
+        sender_sockets = []
+        if not self.unicast:
+            reader_sockets.append(cast(socket.socket, self._listen_socket))
         if self.multi_socket:
             for s in self._respond_sockets:
-                self.engine.add_reader(self.listener, s)
-        # Start the engine only after all
-        # the readers have been added to avoid
-        # missing any packets that are on the wire
-        self.engine.start()
+                reader_sockets.append(s)
+                sender_sockets.append(s)
+
+        for s in reader_sockets:
+            try:
+                transport, protocol = await loop.create_datagram_endpoint(
+                    lambda: AsyncListener(self, loop),
+                    sock=s,
+                )
+                self._readers.append((cast(asyncio.DatagramTransport, transport), protocol))
+                if s in sender_sockets:
+                    self._senders.append((cast(asyncio.DatagramTransport, transport), protocol))
+            except OSError:
+                log.warning("Cannot create endpoint on %s", s)
+
+        for s in self._respond_sockets:
+            if s in reader_sockets:
+                continue
+            try:
+                transport, protocol = await loop.create_datagram_endpoint(
+                    asyncio.DatagramProtocol,
+                    sock=s,
+                )
+                self._senders.append((cast(asyncio.DatagramTransport, transport), protocol))
+            except OSError:
+                log.warning("Cannot create endpoint on %s", s)
+
+    async def async_cache_cleanup(self) -> None:
+        while not self.done:
+            now = current_time_millis()
+            async with self.async_update_records(now, list(self.cache.expire(now))):
+                pass
+            await asyncio.sleep(_CACHE_CLEANUP_INTERVAL)
```

Comment:
Probably a new object for this

Reply:
Actually makes more sense to move it into the Engine class and rename AsyncEngine